core/src/main/java/com/linecorp/armeria/internal/common/KeepAliveHandler.java

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.common;
+
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+
+import javax.annotation.Nullable;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Stopwatch;
+
+import com.linecorp.armeria.common.util.Exceptions;
+
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelFutureListener;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.EventLoop;
+
+/**
+ * A {@link IdleTimeoutScheduler} that writes a PING when neither read nor write was performed for
+ * the specified {@code pingIntervalMillis}, and closes the connection
+ * when neither read nor write was performed within the given {@code idleTimeoutMillis}.
+ */
+public abstract class KeepAliveHandler extends IdleTimeoutScheduler {
```

Comment:
How about merging the two classes into one given that `IdleTimeoutScheduler` is always used together with `KeepAliveHandler`?

Reply:
That sounds good. 👍 It can simplify the hierarchy.